### ISSUE TITLE
ForAll - bug

### ISSUE THREAD

**Issue by zenna** [dreal/dreal3#139]
The following example is to see if we can approximate a function f with a polynomial g.  In this case f is a polynomial so it should be satisfiable

```julia
using DReal
f(x) = x*2
θ = [Var(Float64, -10.0, 10.0) for i = 1:4]
g(x) = θ[1] + θ[2]*x + θ[3]*x^2 + θ[4]*x^3

x = ForallVar(Float64, -10.0, 10.0)
constraint1 = g(x) == f(x) # This constraint works (i.e. SAT with valid model)
constraint2 = abs(g(x) - f(x)) < 0.5 # However, this (incorrectly) is UNSAT
add!(constraint1)
@show is_satisfiable()
result = model(θ...)
@show result
```

In other words constraint2 is a weaker constraint than 1 but it fails.

In addition the old error where calling check() two times in a row can result in different results, has returned.

**Comment by soonhokong** [comment]
1. @zenna, could you check the dReal version that you're using (commit hash)? With the latest version, I have "delta-sat" for both of constraint1 and constraint2.

2. 

> In addition the old error where calling check() two times in a row can result in different results, has returned.

You mean, both of them returns delta-sat answers but with different models, right? If so, it is the expected result. We randomly pick a value from y-space. Also, when we do branching, we randomly select the box to visit among two split boxes.

**Comment by soonhokong** [comment]
@zenna, here are the SMT2 files that I used to test dReal:

 - [constraint1](https://github.com/dreal/dreal3/blob/f205695bfcb14fcee2a47fa32ca090e5812068cf/src/tests/nra/exist_forall_zenna_02.smt2)
 - [constraint2](https://github.com/dreal/dreal3/blob/f205695bfcb14fcee2a47fa32ca090e5812068cf/src/tests/nra/exist_forall_zenna_03.smt2)

**Comment by soonhokong** [comment]
I'll pre-release the latest version in 10min. Please try that one and let me know if the problem is still there.

**Comment by soonhokong** [comment]
> You mean, both of them returns delta-sat answers but with different models, right? If so, it is the expected result. We randomly pick a value from y-space. Also, when we do branching, we randomly select the box to visit among two split boxes.

Obviously, if you use smaller precision, you'll get a smaller difference in models you get from two checks. But it will increase the running time..

**Comment by zenna** [comment]
I'll try the new release.

> You mean, both of them returns delta-sat answers but with different models, right? If so, it is the expected result. We randomly pick a value from y-space. Also, when we do branching, we randomly select the box to visit among two split bo

No that's not what I meant.

I mean:
```julia
...
is_satisfiable() # = true
is_satisfiable() # = false
```

We had this problem before, remember? You fixed it.  The issue was with the precision.

**Comment by soonhokong** [comment]
@zenna, I see. I'll check that.

The prerelease is out. https://github.com/dreal/dreal3/releases/tag/v3.15.07.01 Could you try whether check for constraint2 gives delta-sat with this version?

**Comment by zenna** [comment]
Actually I am using libs built from source in commit 4434cfd3305f4971ca7532488b52ca8d88b32bfd

**Comment by soonhokong** [comment]
@zenna, I see. Maybe SMT2 files are not good for testing this. I'll write C programs testing the problem.

**Comment by zenna** [comment]
With the new binaries I get SAT for both constraints as you did.

**Comment by soonhokong** [comment]
Well... that's good news but I don't understand what makes the change.

I guess the double check problem is still there, could you check with the new binaries?

**Comment by soonhokong** [comment]
> I guess the double check problem is still there, could you check with the new binaries?

I can reproduce the problem. I'll fix it.

**Comment by soonhokong** [comment]
After dReal processes a 'check' command, it keeps the new bound. For the constraint2 example, the initial bounds are:

```
t1 : [-10, 10] = [-10, 10]
t2 : [-10, 10] = [-10, 10]
t3 : [-10, 10] = [-10, 10]
t4 : [-10, 10] = [-10, 10]
```

after the first check, the assignment is:

```
t1 : [-10, 10] = [6.603051493196515, 6.603711864212612]
t2 : [-10, 10] = [-1.184013415037737, -1.183153441737301]
t3 : [-10, 10] = [0.8729532561564891, 0.8737960244086094]
t4 : [-10, 10] = [0.02144443467549646, 0.02156432454809259]
```

When the second check begins, it starts with this "already-pruned" search space. At this point, you may think that "Well, this box is already smaller enough. Why doesn't it return delta-sat immediately?". In current pruning, we always apply pruning operators at least once and then check the width of current box to decide whether to stop or not.

In this particular example, the pruning triggered by the second "check" command makes the box empty. That's why we got UNSAT.

In short, what we have is:

 - First check finds a delta-sat answer.
 - Second check shows that the first delta-sat answer is actually infeasible.

It's bad because in the second check, we already lost the backtracking stack which was maintained in the first check.

A workaround is to wrap each check command with "push" and "pop". For example,

```
;; first check
(push 1)
(check-sat)
(pop 1)

;; second check
(push 1)
(check-sat)
(pop 1)
```

The first pop command restore the search space to the initial one. This should gives the same delta-sat answer (with different model of course).
